substrate/pkg/controller/substrate/cluster/instanceprofile.go

```diff
@@ -53,103 +53,169 @@ var PolicyDocument = aws.String(`{
 				"s3:GetObject",
 				"s3:ListBucket",
 				"ec2:DescribeAddresses",
-				"ec2:AssociateAddress"
+				"ec2:AssociateAddress",
+				"ec2:CreateLaunchTemplate",
+				"ec2:CreateFleet",
+				"ec2:RunInstances",
+				"ec2:CreateTags",
+				"iam:PassRole",
+				"ec2:TerminateInstances",
+				"ec2:DescribeLaunchTemplates",
+				"ec2:DescribeInstances",
+				"ec2:DescribeSecurityGroups",
+				"ec2:DescribeSubnets",
+				"ec2:DescribeInstanceTypes",
+				"ec2:DescribeInstanceTypeOfferings",
+				"ec2:DescribeAvailabilityZones",
+				"ssm:GetParameter"
 			],
 			"Resource": ["*"]
 		}
 	]
 }`)
 
-var ManagedPolicies = []string{
+var ManagedPoliciesForSubstrateNode = []string{
 	"arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
+	"arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
+	"arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy",
+}
+
+var ManagedPoliciesForNodeProvisionedByKarpenter = []string{
+	"arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
+	"arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
+	"arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy",
+	"arn:aws:iam::aws:policy/AmazonEKSClusterPolicy",
+	"arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy",
+}
+
+type roleInfo struct {
+	objectName      *string
+	policyDocument  *string
+	managedPolicies []string
 }
 
 func (i *InstanceProfile) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	for _, desired := range desiredRolesFor(substrate) {
```

Comment:
Are we going to have more than two roles? I think the substrate has: control plane role, data plane role. I'm not sure we'll grow beyond this. Rather than a new type, struct indirection, and looping, it might be more straight forward to just explicitly create both of them.

Reply:
There is one more role we will need to create for guest dataplane nodes.